### ISSUE TITLE
Duplicate fc_id's are often handled wrong in NT02 (and NT03) loading scripts

### ISSUE THREAD

**Issue by edwardsmarc** [edwardsmarc/CASFRI#580]
NT02 and NT03 have many duplicate fc_id's because there is a grid that intersects many polygons, splitting them into smaller polygons that each retain the original fc_id.

NT02 loading script identifies duplicate fc_id's and merges them during loading, summing the areas and selecting attributes assuming that all attributes are duplicated.

This only works for Case 1 outlined below:

### Case 1 - e.g. NT02 fc_id 7002836
- Original polygon is intersected by grid.
- Resulting polygons occur on both sides of the intersect.
- Each polygon is assigned it's area **after** the intersect (i.e. it's true area)
- All attributes except area and geometry are duplicated. 2747 rows:
```
WITH dup AS (
  SELECT fc_id::int
  FROM rawfri.nt03
  WHERE fc_id NOT IN (
	  SELECT fc_id::int
	  FROM rawfri.nt03
	  GROUP BY fc_id::int, wkb_geometry
	  HAVING count(*) > 1
	  UNION ALL
	  SELECT fc_id::int
	  FROM rawfri.nt03
	  GROUP BY fc_id::int, areaha
	  HAVING count(*) > 1)
  GROUP BY fc_id::int
  HAVING count(*) > 1
)
SELECT *
FROM dup a, rawfri.nt03 b 
WHERE a.fc_id::int = b.fc_id::int
```
- In this case, merging and summing the areas is the correct choice. However we could also just add our own unique id and translate them as separate polygons (same method as case 2).

### Case 2 - e.g. NT02 fc_id 1027515
- Original polygon is intersected by grid.
- Resulting polygons occur on the same side of the intersect.
- Each polygon is assigned the same area value, likely the combined area of the individual polygons.
- All attributes except geometry are duplicated.
- This effects 331 rows:
```
WITH dup AS (
  SELECT fc_id::int
  FROM rawfri.nt03
  WHERE fc_id NOT IN (
	  SELECT fc_id::int
	  FROM rawfri.nt03
	  GROUP BY fc_id::int, wkb_geometry
	  HAVING count(*) > 1)
  GROUP BY fc_id::int, areaha
  HAVING count(*) > 1
)
SELECT *
FROM dup a, rawfri.nt03 b 
WHERE a.fc_id::int = b.fc_id::int
```
- In this case the polygons are spatially distinct, any matching polygons on the other side of the grid have a different fc_id. The correct way to process these is as different polygons. The source_area will just be incorrect. We would have to add our own unique id to process these.

### Case 3 - NT03 fc_id 1003807
- NT03 has some polygons that are duplicated.
- Sometimes there are multiple versions of the polygon duplicated, e.g. a disturbance record and forest record, both duplicated (e.g. 1003807).
- Or an empty row of NULLs and zeros, and a forest polygon, both duplicated (e.g. 1019505).
- And cases where polygons that have been split by a grid, are also duplicated. Leading to multiple duplicated polygons all with the same fc_id (e.g. 4000264).
- This problem occurs in 102 polygons. They all have reference years of 1960 or 1970.
```
WITH dup AS (
  SELECT fc_id::int, wkb_geometry
  FROM rawfri.nt03
  GROUP BY fc_id::int, wkb_geometry
  HAVING count(*) > 1
)
SELECT *
FROM dup a, rawfri.nt03 b 
WHERE a.fc_id::int = b.fc_id::int
```
- The only way to process these would be to merge the polygons that have matching attributes and geometries. We would then be left with the distinct versions of the duplicated polygons which could be translated separately. Any with all nulls would not be translated. DST, NFL and LYR rows would be translated as normal.

### Solutions
- Case 3 does not occur in NT02. Easiest fix for NT02 would be not to merge anything, just process polygons as they are in the source data, and add out own poly_id during loading.
- For NT03 we would at least need to address case 3 by merging or deleting duplicated polygons.

**Comment by pedrogit** [comment]
In NT02, what are you trying to fix? The wrong areas for cases 2? Your fix, of not merging them, does not fix the wrong area since both polygons have the combined area. Do what does your fix actually fix?

**Comment by edwardsmarc** [comment]
There are 2 problems with case 2 right now:
1. SRC_INV_AREA is wrong.
2. Polygons are being merged that are not spatially contiguous. e.g. fc_id 1027515 is a merged into a single CASFRI polygon
![image](https://user-images.githubusercontent.com/10728298/107066905-678bca00-6793-11eb-9801-9b97e92c8ee7.png)

**Comment by pedrogit** [comment]
If we merge them together geometrically the SRC_INV_AREA becomes right. No?

Who said polygons had to be spatially continuous to represent a stand or anything else? I'm sure there are thousands of stand spanning two or more spatially discontinous polygons in the rest of the CASFRI coverage.

**Comment by edwardsmarc** [comment]
We'd have to check the summed true areas match the src_inv_area. I was assuming they did but I haven't checked.

It's true they don't need to be contiguous. But I also don't understand why we are putting so much effort into merging them. Why not leave them separate? Is it because we don't want to add our own unique id?

**Comment by pedrogit** [comment]
For some reasons:

1) Yes it is always preferable to use the inventory unique identifier if one exists. But it must be unique. In this case it was not but it was (apparently) easy to fix until you found this issue. We can still fix it.

2) Not trying to fix them would result in some splitted values to become wrong (e.g. SRC_INV_AREA, maybe STRUCTURE_PER? others?)

It true that there is not a strong case for trying to fix them. I guess if we could demonstrate that we can generate a unique identifier for each inventory (sometimes by fixing them like here) and simplify CAS_ID then the case would become stronger.

**Comment by edwardsmarc** [comment]
Note on DISTINCT vs GROUP BY:

- SELECT DISTINCT when you want to keep only one set of values from the duplicates and GROUP BY when you want to aggregate the values together in some way
- Always first ask yourself if you need to aggregate first
- For example to fix case 2 you don't need to aggregate because you don't need to sum the areas. In this case you want to select only one when geometry and area are identical so you do: Use DISTINCT ON (geometry, area). Normally, if all the attributes are NOT all the same, you should use ORDER BY to control which one is taken
- But for case 1 you have to.
- So DISTINCT should be the way to go for case 2 and GROUP BY for case 1

**Comment by edwardsmarc** [comment]
Update:
Case 3 does occur in NT02, we just haden't noticed it because the attributes and geometries were provided separately and we removed duplicates from the attributes assuming all values were the same in duplicate fc_id rows. This is not the case. Which explains the case 3 polygons in NT03. They are many-to-one joins between the duplicated attributes and the polygons.

My proposal for processing NT02 is therefore to do a many-to-one join and create new duplicated polygons for these cases. Then process as per NT03.

But first we need to hear from @MelinaHoule about whether we need both NT02 and NT03. The inventory_list.csv suggests they are the same data with some errors fixed. If this is true, surely we only need NT03?

**Comment by MelinaHoule** [comment]
NT02 and NT03 are the same inventory (#491 ). NT02 never went public. We got it from our contact as soon as they finished assembling the data. The source data didn't reach the validation process on their hand before they shared it with us. There were some issues on species abbreviation. We contacted NT to understand the issue and they offered to send back the latest version of the inventory with error fixed. Here is the answer we got:

_Since the previous inventory was shared, we’ve gone through and fixed many errors that were in the database.  I’ve uploaded the latest version of FVI data..._

We don't need NT02 anymore if we use NT03.
